unmanaged federation technote draft (#149)
apollographql/docs: src/content/technotes/TN0006-unmanaged-federation.mdx

---

> title: TN0006 - Using Unmanaged Federation

Reviewer: (And we use sentence casing for basically all titles and headings)

id: TN0006
tags: [federation, gateway]
---

> [Managed Federation]([https://www.apollographql.com/docs/federation/managed-federation/overview/](https://www.apollographql.com/docs/federation/managed-federation/overview/)) is, in short, a way to run a federated graph utilizing Apollo Studio to provide the necessary supergraph file to your Apollo Gateway or Router using [Apollo Uplink](https://www.apollographql.com/docs/federation/managed-federation/uplink/). Unmanaged Federation, on the other hand, is running a federated graph without the use of [Apollo Uplink](https://www.apollographql.com/docs/federation/managed-federation/uplink/), which requires additional work to support effectively. If you are considering running an unmanaged federated graph, there are a few considerations to keep in mind when doing so.

Reviewer: ```suggestion
With [managed federation](/federation/managed-federation/overview/), your federated supergraph's gateway or router automatically fetches its composed supergraph schema from Apollo. If your supergraph _doesn't_ use managed federation, you need to compose the supergraph schema and provide it to your gateway via another mechanism. This article describes options and considerations for that process.

Reviewer: We should refer to the "supergraph schema file" or "supergraph schema" as opposed to just "supergraph file" to avoid ambiguity with some other supergraph-related files (like Rover composition config)

Author: Got it!

> ### Graph Composition

Reviewer: Moving all these headings up a level so they show up in the article TOC

> There are a few ways to compose your supergrpah that will allow you to use it with your Apollo Gateway or Router, two of which are discussed below. 

Reviewer: ```suggestion
Options for composing your supergraph schema _besides_ managed federation are described below.
```

> #### Build Status Webhook

Reviewer: ```suggestion
### Build status webhook
```

> This approach uses the [Build Status Webhook](https://www.apollographql.com/docs/studio/build-status-notification) to trigger a redeployment of a Gateway or Router by updating on successful builds with a pre-composed supergraph file provided by Apollo Studio. The payload of the webhook contains a URL to fetch the supergraph file necessary, and can be used to download the supergraph file. 

Reviewer: ```suggestion
This approach uses the [build status webhook](/studio/build-status-notification) to trigger redeployment of a gateway or router by updating on successful builds with a composed supergraph schema provided by Apollo Studio. The payload of the webhook contains a URL to download the composed schema. 
```

- Leveraging `rover subgraph introspect` to pull the schema via introspection
- Pulling the schema from a version control service, such as Github or Gitlab

As with the above [Build Status Webhook section](#build-status-webhook), you should be cognizant to use [Rover to run a schema check as part of a CI/CD pipeline](https://www.apollographql.com/docs/rover/commands/subgraphs/#checking-subgraph-schema-changes) or during build time to ensure the changes are not going to break clients. 

For more information on how to accomplish this, see: 

- [https://www.apollographql.com/docs/rover/commands/supergraphs#supergraph-compose](https://www.apollographql.com/docs/rover/commands/supergraphs#supergraph-compose)
- [https://www.apollographql.com/docs/federation/quickstart/local-composition](https://www.apollographql.com/docs/federation/quickstart/local-composition)

### Using the Supergraph File

Similar to getting the supergraph composed, there are a number of ways to approach updating the supergraph in your Gateway or Router. Three options are discussed below. 

> #### Restarting Gateway/Router

Reviewer: Despite being the most straightforward, I recommend putting this option at the bottom since we don't want people to pick it

> #### Restarting Gateway/Router

Author: Will do!

> The simplest option is to simply restart your Gateway or Router processes. If you passed the file into the instances via a build process (such as a Docker image), you can redeploy to take effect. This isn't recommended as there are ways to avoid any redeployments, discussed below.

Reviewer: ```suggestion
The most straightforward option is to restart all of your gateway or router instances. If you pass your supergraph schema file to these instances via a build process (such as a Docker image), you can redeploy for your changes to take effect.

This method is _not_ recommended, because the restart causes downtime for your supergraph. The methods described above can update your supergraph schema _without_ requiring redeployment.
```

> #### Hot Reloading

Reviewer: ```suggestion
### Hot reloading
```

> **If you are using the Router, you can use the `--hr` flag to enable hot reloading natively; no further configuration is needed**

Reviewer: ```suggestion
**If you're using the Apollo Router,** you can enable hot reloading by passing the `--hr` option to the `router` executable. No other configuration is required.
```

> The flip of restarting processes in Gateway is to enable hot reloading the file using a simple watcher. 

Reviewer: **If you're using the `@apollo/gateway` library,** you can enable hot reloading by setting up a basic watcher.

> For example, if you're using Docker you can pass in the file as a mounted volume such as the below where the local supergraph file is mounted to `/etc/config/supergraph.graphql` for the Gateway to use. 

Reviewer: ```suggestion
For example, if you're using Docker, you can pass the supergraph schema file as a mounted volume. In the following example, the supergraph schema is mounted to `/etc/config/supergraph.graphql`:
```

```Dockerfile
version: '3'
services:
  apollo-gateway:
    container_name: apollo-gateway
    build: ./gateway
    volumes:
      - ./supergraph.graphql:/etc/config/supergraph.graphql
    ports:
```

> Once you've passed the file to the instances, you'll want to make sure to use the `supergraphSdl` option in Gateway to hot reload the graph. An example of this would be below, which uses `chokidar` to watch for file updates. 

Reviewer: ```suggestion
After you make the schema file available to your gateway instances, pass a function to the `supergraphSdl` option of your `ApolloGateway` constructor that enables the gateway to hot reload the schema. The example function below uses the `chokidar` library to watch for file updates:
```

const { ApolloGateway } = require('@apollo/gateway');
const { readFile } = require('fs/promises');
const chokidar = require('chokidar');
const supergraphSdlLocation = '/etc/config/supergraph.graphql';

const gateway = new ApolloGateway({
  async supergraphSdl({ update, healthCheck }) {
    const watcher = chokidar.watch(supergraphSdlLocation);
    watcher.on('change', async ()=>{
      try {
          const updatedSupergraph = await readFile(supergraphSdlLocation, 'utf-8');
          // optional health check update to ensure our services are responsive
          await healthCheck(updatedSupergraph);
          // update the supergraph schema
          update(updatedSupergraph);
        } catch (e) {
          // handle errors that occur during health check or while updating the supergraph schema
          console.error(e);
        }
    });
    return {
      supergraphSdl: await readFile(supergraphSdlLocation, 'utf-8'),
      async cleanup() {
        await watcher.close();
      }
    }
  },
});

const server = new ApolloServer({
  gateway,
});
server.listen().then(({ url }) => {
  console.log(`🚀 Server ready at ${url}`);
});
```

> For more information on this approach, we recommend [checking out the Gateway docs to learn more](https://github.com/apollographql/federation/blob/main/docs/source/gateway.mdx#updating-the-supergraph-schema) including further steps to harden the implementation. 

Reviewer: ```suggestion
For more information on this approach, [see the Apollo Gateway docs](/federation/gateway.mdx#updating-the-supergraph-schema).
```

> #### Custom SupergraphManager

Reviewer: ```suggestion
### Custom `SupergraphManager`
```

> Lastly, if you have complex logic that would be better suited for a seperate class (for example, listening to an AWS Simple Notification Service (SNS) topic), you can create a custom implementation of the `SupergraphManager` class to be passed to the `supergraphSdl` option described earlier. 

Reviewer: ```suggestion
If you have complex logic that's be better suited to its own class (for example, listening to an AWS Simple Notification Service (SNS) topic), you can create a custom implementation of the `SupergraphManager` class to pass to the `supergraphSdl` option.
```

> This option provides the most customizability, but comes at the cost of writing additional code- however the interface for writing the class is extremly simple. For an example, [see the Uplink implementation of `SupergraphManager` on how to do so](https://github.com/apollographql/federation/blob/main/gateway-js/src/supergraphManagers/UplinkSupergraphManager/index.ts). 

Reviewer: ```suggestion
This option provides the most customizability, but it also requires writing additional code. For an example, [see the Uplink implementation of `SupergraphManager`](https://github.com/apollographql/federation/blob/main/gateway-js/src/supergraphManagers/UplinkSupergraphManager/index.ts). 
```

import * as chokidar from 'chokidar';
import {readFile} from 'fs/promises';

export class FileWatcherManager implements SupergraphManager {
    private filename: string; 

    constructor({
        filename,
    }:{
        filename: string
    }){
        this.filename = filename
    }
    
    public async initialize({ update, healthCheck }: SupergraphSdlHookOptions) {
        let watcher = chokidar.watch(this.filename)

        watcher.on('change', async ()=>{
            try {
                const updatedSupergraph = await readFile(this.filename, 'utf-8');
                // optional health check update to ensure our services are responsive
                await healthCheck(updatedSupergraph);
                // update the supergraph schema
                update(updatedSupergraph);
            } catch (e) {
                // handle errors that occur during health check or while updating the supergraph schema
                console.error(e);
            }
        });
        
        return {
            supergraphSdl: await readFile(this.filename, 'utf-8'),
            cleanup: async () => {
                await watcher.close()
            },
          };
    }   
}
```

> And to use it, you will instantiate it for the `sueprgraphSdl` property. For example:

Reviewer: ```suggestion
To use your custom `SupergraphManager`, you instantiate it and pass it as the `sueprgraphSdl` property, like so:

```

```ts
const gateway = new ApolloGateway({
    supergraphSdl: new FileWatcherManager({filename: './supergraph.graphql'})
})
```

### Restart the gateway/router (not recommended)

> The most straightforward option is to restart all of your gateway or router instances. If you pass your supergraph schema file to these instances via a build process (such as a Docker image), you can redeploy for your changes to take effect.

Reviewer: ```suggestion
The most straightforward option for updating your supergraph schema to restart all of your gateway or router instances. If you pass your supergraph schema file to these instances via a build process (such as a Docker image), you can redeploy for your changes to take effect.
```

> This method is _not_ recommended, because the restart causes downtime for your supergraph. The methods described above can update your supergraph schema _without_ requiring redeployment.

Reviewer: ```suggestion
**This method is _not_ recommended,** because the restart causes downtime for your supergraph. The methods described above can update your supergraph schema _without_ requiring redeployment.
```

## Composing the supergraph schema

Options for composing your supergraph schema _besides_ managed federation are described below.

### Build status webhook

This approach uses the [build status webhook](/studio/build-status-notification) to trigger redeployment of a gateway or router by updating on successful builds with a composed supergraph schema provided by Apollo Studio. The payload of the webhook contains a URL to download the composed schema. 

The advantage of this approach is that you have a confirmed build that composes successfully. However, the webhook notification is sent _before_ operation checks are run against your historical operation data. This means you should [run schema checks with Rover as part of your CI/CD pipeline](/rover/commands/subgraphs/#checking-subgraph-schema-changes).

> ### Local composition

Author: Got it, I can add a note that this won't apply to contracts.

                await healthCheck(updatedSupergraph);
                // update the supergraph schema
                update(updatedSupergraph);
            } catch (e) {
                // handle errors that occur during health check or while updating the supergraph schema
                console.error(e);
            }
        });
        
        return {
            supergraphSdl: await readFile(this.filename, 'utf-8'),
            cleanup: async () => {
                await watcher.close()
            },
          };
    }   
}
```

To use your custom `SupergraphManager`, you instantiate it and pass it as the `sueprgraphSdl` property, like so:

```ts
const gateway = new ApolloGateway({
    supergraphSdl: new FileWatcherManager({filename: './supergraph.graphql'})
})
```

### Restart the gateway/router (not recommended)

The most straightforward option for updating your supergraph schema is to restart all of your gateway or router instances. If you pass your supergraph schema file to these instances via a build process (such as a Docker image), you can redeploy for your changes to take effect.

> **This method is _not_ recommended,** because the restart causes downtime for your supergraph. The methods described above can update your supergraph schema _without_ requiring redeployment.

Reviewer 2: ```suggestion
**This method is _not_ recommended,** because the restart can potentially cause downtime for your supergraph. The methods described above can update your supergraph schema _without_ requiring redeployment.
```

If there is a slow traffic cut over with blue/green deploy the could be a no-down-time deploy

### Graph Composition

When running Unmanaged Federation, Apollo no longer builds your supergraph for you; instead, you'll need to leverage [Rover]([https://www.apollographql.com/docs/rover/](https://www.apollographql.com/docs/rover/)) to build the supergraph as part of a larger CI/CD pipeline in order to then be used by your Apollo Gateway or Router. 

The process for setting this up can vary depending on CI tooling, but you'll need to leverage `rover supergraph compose` to build the supergraph GraphQL file from a configuration file.

To run the `rover supergraph compose` command, you'll need access to each subgraph's schema SDL. There are a few ways to do this: 

- Leveraging `rover subgraph introspect` to pull the schema via introspection
- Pulling the schema from a version control service, such as Github or Gitlab

> Once you have the SDL for each subgraph you run, you will then need to update the supergraph file on your Apollo Gateway or Router via a deployment and restart. 

Reviewer 3: you will also have to do this on every subgraph schema change.
